src/Multi_filtration/include/gudhi/Multi_critical_filtration.h

```diff
@@ -37,400 +40,573 @@ namespace Gudhi::multi_filtration {
+      }
+      return projection;
+    } else {
+      U projection = std::numeric_limits<U>::max();
+      for (const auto &y : f) {
+        projection = std::min(projection, compute_linear_projection(y, x));
       }
+      return projection;
     }
-    if constexpr (verbose) std::cout << "x: " << x << " is useful, removing unnecessary entries" << std::endl;
-    multi_filtration_.erase(std::remove_if(multi_filtration_.begin(), multi_filtration_.end(),
-                                           [&x](const One_critical_filtration<T> &y) {
-                                             if constexpr (verbose) {
-                                               if (dominates(y, x)) {
-                                                 std::cout << "Removing y=" << y << std::endl;
-                                               } else {
-                                                 std::cout << "Keeping y=" << y << std::endl;
-                                               }
-                                             }
-                                             return dominates(y, x);
-                                           }),
-                            multi_filtration_.end());
+  }
 
-    multi_filtration_.push_back(x);
+  /*
+   * Same as its one critical counterpart.
+   * If a grid is given, projects multi_filtration_ on this grid and returns
+   * multi critical filtration composed of the coordinates in the given grid
+   *
+   */
+  template <typename out_type = std::int32_t, typename U = T>
+  friend Multi_critical_filtration<out_type> compute_coordinates_in_grid(const Multi_critical_filtration &f,
+                                                                         const std::vector<std::vector<U>> &grid)
+  {
+    Multi_critical_filtration<out_type> coords = f.as_type<out_type>();
+    coords.project_onto_grid(grid);
+    return coords;
   }
-  inline void re_clean() {
-    // Ensures all points are useful again. Can be useful if points are added manually.
-    // TODO : maybe optimize
-    Multi_critical_filtration<value_type> out;  // should be inf
-    out.multi_filtration_.reserve(this->multi_filtration_.size());
-    for (const auto &x : multi_filtration_) {
-      out.add_point(x);
+
+  /*
+   * Same method as the one in OneCriticalFiltration.
+   * Given a grid, and assuming that `this` are the coordinates
+   * in this grid, evaluate `this` into this grid
+   */
+  template <typename U>
+  friend Multi_critical_filtration<U> evaluate_coordinates_in_grid(const Multi_critical_filtration &f,
+                                                                   const std::vector<std::vector<U>> &grid)
+  {
+    Multi_critical_filtration<U> out;
+    out.set_num_generators(f.num_generators());
+    for (std::size_t i = 0; i < f.num_generators(); ++i) {
+      out[i] = evaluate_coordinates_in_grid(f[i], grid);
     }
-    std::swap(multi_filtration_, out);
+    out.simplify();
+    return out;
```

Comment:
I am pretty sure that `evaluate_coordinates_in_grid` didn't used add point before neither? I am not sure what you mean? The only thing I changed here is adding the simplified method. And using `set_num_generators` because `Multi_critical_filtration<U> out(this->num_generators());` is not doing what you want here.

Reply:
ah ok